lib/url.c

```diff
@@ -5256,6 +5293,97 @@ static CURLcode set_login(struct connectdata *conn,
   return result;
 }
 
+/*
+ * Extracts the host name and the port from the "connect to host" string.
+ */
+static CURLcode parse_connect_to_host(struct SessionHandle *data,
+                            struct connectdata *conn, char *conn_to_host)
+{
+  char *conn_to_host_dup = NULL;
+  long port = 0;
+  char *host_portno;
+
+  char *hostptr;
+  char *portptr;
+
+  if(!conn_to_host || !*conn_to_host)
+    return CURLE_OK;
+
+  conn_to_host_dup = strdup(conn_to_host);
+  if(!conn_to_host_dup)
+    return CURLE_OUT_OF_MEMORY;
+
+  hostptr = conn_to_host_dup;
+
+  /* start scanning for port number at this point */
+  portptr = hostptr;
+
+  /* detect and extract RFC6874-style IPv6-addresses */
+  if(*hostptr == '[') {
+    char *ptr = ++hostptr; /* advance beyond the initial bracket */
+    while(*ptr && (ISXDIGIT(*ptr) || (*ptr == ':') || (*ptr == '.')))
+      ptr++;
+    if(*ptr == '%') {
+      /* There might be a zone identifier */
+      if(strncmp("%25", ptr, 3))
+        infof(data, "Please URL encode %% as %%25, see RFC 6874.\n");
+      ptr++;
+      /* Allow unresered characters as defined in RFC 3986 */
+      while(*ptr && (ISALPHA(*ptr) || ISXDIGIT(*ptr) || (*ptr == '-') ||
+                     (*ptr == '.') || (*ptr == '_') || (*ptr == '~')))
+        ptr++;
+    }
+    if(*ptr == ']')
+      /* yeps, it ended nicely with a bracket as well */
+      *ptr++ = 0;
+    else
+      infof(data, "Invalid IPv6 address format\n");
+    portptr = ptr;
+    /* Note that if this didn't end with a bracket, we still advanced the
+     * hostptr first, but I can't see anything wrong with that as no host
+     * name nor a numeric can legally start with a bracket.
+     */
+  }
+
+  /* Get port number off server.com:1080 */
+  host_portno = strchr(portptr, ':');
+  if(host_portno) {
+    char *endp = NULL;
+    *host_portno = 0x0; /* cut off number from host name */
+    host_portno ++;
+    port = strtol(host_portno, &endp, 10);
+    if((endp && *endp) || port >= 65536) {
+      infof(data, "No valid port number in connect to host string (%s)\n",
+        host_portno);
+      port = 0;
+    }
+  }
+
+  if(!*hostptr)
+    infof(data, "Host name is missing in connect to host string\n");
```

Comment:
> shouldn't illegal syntax cause this function to return an error?

Currently, this function is very similar to `parse_proxy()`. It emits error messages but does not cancel the request.